docs/topics/addons.rst

```diff
@@ -0,0 +1,348 @@
+either an add-on or another add-on path.
+
+
+Add-on base class
+=================
+
+Scrapy comes with a built-in base class for add-ons which provides some
+convenience functionality:
+
+* basic settings can be exported via :meth:`~scrapy.addons.Addon.export_basics`,
+  configurable via :attr:`~scrapy.addons.Addon.basic_settings`.
+* a single component (e.g. an item pipeline or a downloader middleware) can be
+  inserted into Scrapy's settings via
+  :meth:`~scrapy.addons.Addon.export_component`, configurable via
+  :attr:`~scrapy.addons.Addon.component_type`,
+  :attr:`~scrapy.addons.Addon.component_key`,
+  :attr:`~scrapy.addons.Addon.component`, and the ``order`` key in
+  :attr:`~scrapy.addons.Addon.default_config`.
+* the add-on configuration can be exposed into Scrapy's settings via
+  :meth:`~scrapy.addons.Addon.export_config`, configurable via
+  :attr:`~scrapy.addons.Addon.default_config`,
+  :attr:`~scrapy.addons.Addon.config_mapping`, and
+  :attr:`~scrapy.addons.Addon.settings_prefix`.
+
+By default, the base add-on class will expose the add-on configuration into
+Scrapy's settings namespace, in caps and with the add-on name prepended. It is
+easy to write your own functionality while still being able to use the
+convenience functions by overwriting
+:meth:`~scrapy.addons.Addon.update_settings`.
+
+.. module:: scrapy.addons
+   :noindex:
+
+.. autoclass:: Addon
+   :members:
+
+
+Add-on examples
+===============
+
+Set some basic configuration using the :class:`Addon` base class::
+
+    from scrapy.addons import Addon
+
+    class MyAddon(Addon):
+        name = 'myaddon'
+        version = '1.0'
+        component = 'path.to.mypipeline'
+        component_type = 'ITEM_PIPELINES'
+        component_order = 200
+        basic_settings = {
+            'DNSCACHE_ENABLED': False,
+        }
+
+Check dependencies::
+
+    from scrapy.addons import Addon
+
+    class MyAddon(Addon):
+        name = 'myaddon'
+        version = '1.0'
+
+        def update_settings(self, config, settings):
+            try:
+                import boto
+            except ImportError:
+                raise RuntimeError("myaddon requires the boto library")
+            self.export_config(config, settings)
+
+Enable a component that lives relative to the add-on (see
+:ref:`topics-api-settings`)::
+
+    from scrapy.addons import Addon
+
+    class MyAddon(Addon):
+        name = 'myaddon'
+        version = '1.0'
+        component = __name__ + '.downloadermw.coolmw'
+        component_type = 'DOWNLOADER_MIDDLEWARES'
+        component_order = 900
+
+Instantiate components ad hoc::
+
+    from path.to.my.pipelines import MySQLPipeline
+
+    class MyAddon(object):
+        name = 'myaddon'
+        version = '1.0'
+
+        def update_settings(self, config, settings):
+            mysqlpl = MySQLPipeline(password=config['password'])
+            settings.set(
+                'ITEM_PIPELINES',
+                {mysqlpl: 200},
```

Comment:
Hm, I don't think this example works at all; ITEM_PIPELINES expects pipeline classes, not pipeline instances as keys.

Reply:
So, the original PR changed load_object() to add support for types and instances. We added support for types in a separate PR but didn't add support for instances. I've removed changes to load_object() from here but this part of the doc still assumes instances are supported.